### ISSUE TITLE
exit code 1 instead of showing logs

### ISSUE THREAD

**Issue by jamrizzi** [drone/drone#2380]
Drone shows the logs until it fails, then it replaces the output of the logs with . . .

```sh
exit code 1
```

It makes it almost impossible to debug my projects when they fail

**Comment by tboerger** [comment]
I can not confirm such a behavior, so you should provide more details to make it possible to reproduce, but beside that I want to kindly ask to create a thread on discourse.drone.io as the issue template already states that the issue tracker should only be used for confirmed bugs and not for support requests.

**Comment by bradrydzewski** [comment]
I am unable to reproduce this behavior, and have not received any similar reports ...

**Comment by saward** [comment]
I have this issue occasionally.  Just googled to find out why, and found this thread.  I retried the same build again with a trivial change to the repo (since cancelling such jobs still doesn't seem to properly cancel them), and it worked that time.  So it seems intermittent.  Maybe it used a different drone agent the second time around.

Certainly seems like a bug to not show the output.  On previous occasions, I think I've seen an error flash up before emptying the logs on the page.  I tried to video it this time, but there wasn't an error so I can't provide you with that information.

**Comment by tboerger** [comment]
All errors are also logged by the agent, so please read your logs and provide more information.

**Comment by saward** [comment]
I've extracted what I think are the right logs from the agent.  If relevant, this is running in kubernetes with the agents using dind.

```
{"time":"2018-04-17T06:04:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"received execution"}
{"time":"2018-04-17T06:04:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"listen for cancel signal"}
{"time":"2018-04-17T06:04:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2018-04-17T06:04:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2018-04-17T06:04:34Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","message":"log stream opened"}
{"time":"2018-04-17T06:04:41Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","message":"log stream copied"}
{"time":"2018-04-17T06:04:41Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","message":"log stream uploading"}
{"time":"2018-04-17T06:04:41Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","message":"log stream upload complete"}
{"time":"2018-04-17T06:04:41Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","message":"log stream closed"}
{"time":"2018-04-17T06:04:42Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":true,"message":"update step status"}
{"time":"2018-04-17T06:04:42Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":true,"message":"update step status complete"}
{"time":"2018-04-17T06:04:42Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2018-04-17T06:04:42Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2018-04-17T06:05:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:05:40Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","message":"log stream opened"}
{"time":"2018-04-17T06:06:33Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","message":"log stream copied"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","message":"log stream uploading"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","message":"log stream upload complete"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","message":"log stream closed"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","exit_code":0,"exited":true,"message":"update step status"}
{"time":"2018-04-17T06:06:43Z","level":"debug","repo":"masked/api","build":"35","id":"3238","image":"golang:1.10","stage":"test","exit_code":0,"exited":true,"message":"update step status complete"}
{"time":"2018-04-17T06:06:44Z","level":"debug","repo":"masked/api","build":"35","id":"3238","error":"","exit_code":0,"message":"pipeline complete"}
{"time":"2018-04-17T06:06:44Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"uploading logs"}
{"time":"2018-04-17T06:06:44Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"uploading logs complete"}
{"time":"2018-04-17T06:06:44Z","level":"debug","repo":"masked/api","build":"35","id":"3238","error":"","exit_code":0,"message":"updating pipeline status"}
{"time":"2018-04-17T06:06:44Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"stop listening for cancel signal"}
{"time":"2018-04-17T06:06:46Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"updating pipeline status complete"}
{"time":"2018-04-17T06:06:46Z","level":"debug","message":"request next execution"}
{"time":"2018-04-17T06:06:46Z","level":"debug","repo":"masked/api","build":"35","id":"3238","message":"pipeline done"}
{"time":"2018-04-17T06:07:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"received execution"}
{"time":"2018-04-17T06:07:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"listen for cancel signal"}
{"time":"2018-04-17T06:07:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2018-04-17T06:07:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2018-04-17T06:07:38Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","message":"log stream opened"}
{"time":"2018-04-17T06:07:44Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","message":"log stream copied"}
{"time":"2018-04-17T06:07:44Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","message":"log stream uploading"}
{"time":"2018-04-17T06:07:44Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","message":"log stream upload complete"}
{"time":"2018-04-17T06:07:44Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","message":"log stream closed"}
{"time":"2018-04-17T06:07:45Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":true,"message":"update step status"}
{"time":"2018-04-17T06:07:45Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/git:latest","stage":"clone","exit_code":0,"exited":true,"message":"update step status complete"}
{"time":"2018-04-17T06:07:45Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2018-04-17T06:07:45Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2018-04-17T06:07:45Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","message":"log stream opened"}
{"time":"2018-04-17T06:08:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","message":"log stream copied"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","message":"log stream uploading"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","message":"log stream upload complete"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","message":"log stream closed"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","exit_code":0,"exited":true,"message":"update step status"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"golang:1.10","stage":"test","exit_code":0,"exited":true,"message":"update step status complete"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/docker:latest","stage":"publish","exit_code":0,"exited":false,"message":"update step status"}
{"time":"2018-04-17T06:08:50Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/docker:latest","stage":"publish","exit_code":0,"exited":false,"message":"update step status complete"}
{"time":"2018-04-17T06:08:51Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/docker:latest","stage":"publish","message":"log stream opened"}
{"time":"2018-04-17T06:09:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:09:38Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/docker:latest","stage":"publish","exit_code":1,"exited":true,"message":"update step status"}
{"time":"2018-04-17T06:09:38Z","level":"debug","repo":"masked/api","build":"36","id":"3241","image":"plugins/docker:latest","stage":"publish","exit_code":1,"exited":true,"message":"update step status complete"}
{"time":"2018-04-17T06:09:39Z","level":"debug","repo":"masked/api","build":"36","id":"3241","error":"","exit_code":1,"message":"pipeline complete"}
{"time":"2018-04-17T06:09:39Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"uploading logs"}
{"time":"2018-04-17T06:10:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:11:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:12:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:13:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:14:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
{"time":"2018-04-17T06:15:37Z","level":"debug","repo":"masked/api","build":"36","id":"3241","message":"pipeline lease renewed"}
```

.drone.yml file is similar to:

```
workspace:
  base: /go

pipeline:
  test:
    image: golang:1.10
    commands:
      - go get -u github.com/golang/dep/cmd/dep
      - dep ensure
      - go test

  publish:
    image: plugins/docker
    repo: quay.io/masked/api
    registry: quay.io
    tags: ${DRONE_BRANCH}-${DRONE_COMMIT_SHA:0:12}
    secrets: [ docker_username, docker_password ]
    when:
      branch: [master, develop, advance]

  publish:
    image: plugins/docker
    repo: quay.io/masked/api
    registry: quay.io
    tags: ${DRONE_BRANCH}-${DRONE_TAG}
    secrets: [ docker_username, docker_password ]
    when:
      event: tag

```

It was the 'publish' step that triggered the displaying of only 'exit code 1'.

**Comment by bradrydzewski** [comment]
It looks like your Docker daemon is hanging. I can see the log stream is opened:

```
..."stage":"publish","message":"log stream opened"}
```

But I do not see a subsequent message indicating the log stream is closed. There have been multiple issues logging in the moby (docker) repository related to logs hanging. I recommend upgrading to the latest Docker version which include a number of fixes.

**Comment by saward** [comment]
Thanks, I'll try that out and report back here if I notice ongoing issues.

**Comment by saward** [comment]
I've caught the error, and it looks like it's not docker related, or not purely.  This is with now running a much newer version of docker and drone versions 0.8.5, but still same  issue.  I recorded my screen as the logs outputted, so I could catch the error before the whole log was replaced with a simple "exit code 1".  Here is the error that showed up very briefly before being no longer visible (typed by hand, so please excuse typos):

```
Step 5/8 : RUN go install bitbucket.org/masked/api
---> Running in d7ecc0e42a0a
# bitbucket.org/masked/api/vendor/github.com/graph-gophers/graphql-go/trace
vendor/github.com/graph-gophers/graphql-go/trace/validation_trace.go:7: syntax error: unexpected = in type declaration
The command '/bin/sh -c go install bitbucket.org/masked/api' returned a non-zero code: 2 time="2018-03-17T10:24:34Z" level=fatal msg="exit status 2"
```

The lines after "---> Running" were all red.

I've now fixed my build (since I could see the error -- upgraded Go version for base image), but it was difficult not being able to see what the error was when it flashes up very briefly before being replaced with 'Exit code 1'.

Let me know if I can provide any more useful info.

**Comment by bradrydzewski** [comment]
not sure I agree with the assessment. That is an error message _inside_ the build container. That will not freeze the host machine docker logs streaming. Did you look at your agent logs to see if there was a lot stream closed entry? If not, it remains a docker issue.

**Comment by fringd** [comment]
I'm seeing this too. Important debug output at the end of a fail is getting dropped.

**Comment by fringd** [comment]
any chance we could re-open this?

**Comment by codejamninja** [comment]
Ok, well if drone isn't the problem, does anyone have the slightest idea why the docker logs would be hanging. If the issue is filed on docker please provide a link. If the issue is not filed on docker, could someone who understands the error file it?

**Comment by bradrydzewski** [comment]
@codejamninja I provided four links above in an earlier comment

**Comment by codejamninja** [comment]
Ah, I didn't notice them. Thanks @bradrydzewski

**Comment by saward** [comment]
From the user's perspective, drone here operates in a way that isn't expected: there's a useful stream of messages, and then it disappears.  I understand that the way drone has been built means its operating as designed.  But in this instance, an upstream failure with docker means that drone doesn't operate how a user would expect.  Without understanding the internal details, I look at the messages that *were* there and wonder why I can no longer see them.

Is docker likely to resolve this issue soon?  If not, and if this issue is likely to recur in the future, would it be worth changing the way drone operates so that it can handle upstream failures like this?

**Comment by bradrydzewski** [comment]
0.9 implements a new runtime engine. If this issue can be reproduced in 0.9 we would be happy to accept a pull request from someone that can reproduce the problem and produce a patch.

**Comment by fringd** [comment]
yeah all those issues are closed a long time ago. I don't think it's enough to say "not my fault" and move on. :/

**Comment by bradrydzewski** [comment]
@fringd if you disagree with my assessment send a patch

**Comment by fringd** [comment]
Sorry I do appreciate the great work people are doing on drone. I haven't got time to learn the whole context of how it works. I'm just here as a user reporting a regression.

**Comment by codejamninja** [comment]
@fringd, @bradrydzewski has a good point. I think the reason the issues were closed a long time ago is because Docker fixed the problem in a new release. If Docker truly is where the problem is coming from, then the only way to fix it is to implement a newer runtime engine, which as stated previously by @bradrydzewski, will be implemented in the 0.9 release of drone.

**Comment by fringd** [comment]
Oh, so the issue is already fixed on master? just not released?

**Comment by saward** [comment]
Will let you know if I see it happening in 0.9 once that's out and I've used it a bit.

**Comment by codejamninja** [comment]
@fringd, I'm guessing so. Time will tell.

**Comment by fringd** [comment]
Okay, well that was the main thing I was uppity about. It's a real bug that's affecting me as a user, and having the ticket closed as not my was frustrating.

Any idea why this started happening a month or so ago when it wasn't a problem before? It's a very consistent bug.

**Comment by codejamninja** [comment]
I have no idea why. I haven't had a chance to really look into. All I did was file the bug.

**Comment by fringd** [comment]
@saward did you try updating to the latest docker version?

**Comment by bradrydzewski** [comment]
> yeah all those issues are closed a long time ago.

One of these issues was closed as recently as Nov '17 which means it would be part of the Docker 18.x release. Most linux distributions are still running 17.X and kubernetes is still running 13.x. So upgrading your Docker daemon to 18.x could help.

>  It's a real bug that's affecting me as a user

Have you considered engaging Docker support? We see from the agent logs that the `docker logs` command blocks infinitely. In 0.9 we make one call to `docker logs` instead of two, which may reduce the frequency in which you encounter this docker issue but may not eliminate it. Any issues need to be fixed upstream.

I stand by my assessment that this appears to be a Docker issue and does not belong in the Drone issue tracker. If you can prove otherwise we can re-open the issue.

**Comment by fringd** [comment]
Okay I'll try updating docker and see if that fixes things 🤞

**Comment by saward** [comment]
@fringd I updated the agent's dind to 18.03.1-ce-rc1-dind, and still had the issue.  I don't know the way Drone work well enough though to know if that's the docker daemon that needs to be updated, or if this output is from a different place.

**Comment by geraldoandradee** [comment]
I'm having this same behaviour. When succeeded all logs is there. When fails I saw only `exit 1` (default – <testname>: Failure).

**Comment by Benjamin-Dobell** [comment]
Just a heads up that I'm seeing this as well in Kubernetes deployment.

The only output from the docker build job is (as in Docker log):

```
{"arch":"amd64","build":3,"level":"info","machine":"ip-[REDACTED].ap-southeast-2.compute.internal","msg":"runner: start execution","os":"linux","pipeline":"default","repo":"[REDACTED]","stage":1,"stage-id":22,"time":"2019-05-01T08:50:46Z"}
{"arch":"amd64","build":3,"error":"build_and_push : exit code 1","level":"info","machine":"ip-[REDACTED].ap-southeast-2.compute.internal","msg":"runner: execution failed","os":"linux","pipeline":"default","repo":"[REDACTED]","stage":1,"stage-id":22,"time":"2019-05-01T08:58:59Z"}
```

As with previous reports, successful logs are accessible, it's just failed builds whereby the logs are going missing - but are visible during build time.

I will of course investigate and report back. Not a Go dev, but assuming I can find and understand the root cause I will submit a (probably poorly implemented) patch.

**Comment by Benjamin-Dobell** [comment]
Okay, I don't know about everyone else, so grain of salt. Particularly since there has been a docker logging issue identified further up in this thread. Therefore, although the symptoms are the same/similar, the underlying cause may different. Nonetheless, the following may be useful for at least _some_ people whose logs are going missing...

My issue with my logs not being persisted has nothing to do with _failed_ logs not being persisted and everything to do with _long_ logs not being persisted.

After a quick dive through Drone internals I can see the runner uploads logs using the `/rpc/v1/uploads` end-point, specifically:

* https://github.com/drone/drone/blob/5a0a07e55352a74d1bdd59f834dbd69616d1b96f/operator/runner/runner.go#L471
* https://github.com/drone/drone/blob/f75ea6d4cda25480536ce86dda109d064dd48bfb/operator/manager/rpc/client.go#L206

Lo and behold, by Nginx logs:

> 2019/05/01 13:43:04 [error] 51#51: *4149446 client intended to send too large body: 1376560 bytes, client: [IP_REDACTED], server: drone.[DOMAIN_REDACTED], request: "POST /rpc/v1/upload?id=78 HTTP/1.1", host: "drone.[DOMAIN_REDACTED]"
> [IP_REDACTED] - [[IP_REDACTED]] - - [01/May/2019:13:43:04 +0000] "POST /rpc/v1/upload?id=78 HTTP/1.1" 413 183 "-" "Go-http-client/1.1" 308 0.007 [-] - - - - 38fb500284ded009ed723974f054add4

Now, Nginx for this k8s cluster is actually configured to allow the uploading of 3D models (max 256mb). I've just gone ahead and double checked that this is indeed working from my browser (outside the cluster). However, that HTTP 413 is obviously being generated for a reason; so something is misconfigured on my end.

TLDR; check your Nginx (or similar) logs.
